File: src/api/subsonic/methods/getAlbum.ts
import {
  handleJsonResponse,
  makeSubsonicRequest,
  parseJsonResponse,
} from '../makeRequest';
import { Album } from '../types/album';

export function subsonicGetAlbum(id: string) {
  return makeSubsonicRequest({ method: 'rest/getAlbum', id })
    .flatMap(response => parseJsonResponse(response, { album: Album }))
    .flatMap(json => handleJsonResponse(json['subsonic-response']))
    .map(result => result.album);
}

File: src/api/subsonic/getAlbum.ml
open Fx
open Monad_syntax
open Utils

let make id =
  let* { response = { status = `ok; album } } =
    let params = Js.Dict.empty () in
    Js.Dict.set params "id" (Value id);
    makeRequest { _method = "rest/getAlbum"; params } Album.of_json
  in
  Ok album
